pna.p4

```diff
@@ -599,6 +599,21 @@ struct pna_main_output_metadata_t {
 // + send_to_port
 
 
+// recirculate() - Cause the packet to be recirculated when it
+// finishes completing the main control.
+//
+// Invoking recirculate() is supported only within the main control.
```

Comment:
Maybe this would be better comment:
```
Invoking recirculate() is supported directly or indirectly called from the main control.
```

I know this is how it is written in other locations but being more explict of what it means be invoked from the main control is always better.

Reply:
It would be good to also specify precisely what intrinsic metadata changes, and if that is agreed upon behavior between all architectures.

Reply:
I would like to explore following types of recirculations -
1. parser -> control -> deparser -- recirc1 --> parser ..... (as shown in pna arch diagram) - User metadata is NOT preserved. Any data needed is added as recirc-header, which will be extracted by parser.
2. parser -> control -> -- recirc2 --> control .... -> deparser - User metadata is preserved.

recirc1 and recirc2 (to be named appropriately) are both invoked in action routines and take effect at the end of the pipeline.

Reply:
With the 2nd form of doing recirc, maybe I suggest, the name redocontrol or something similar to that.
With that being said my hardware can support the 2nd form though I am not so sure the 2nd form is useful; can you think of an usecase for each form?
I can think of removing the xvlan and doing some ACLs and then recirculating it and redoing the parsing to get a new set of headers. That would be the first form.
But I can't think of use case for the 2nd form though unless we have the original parser also get the inner headers. I am not 100% sure if that will always work.

Reply:
Use cases of 2nd form I can think of -
- implementing longer (stages) pipeline on a target that has smaller number of stages. Each pass launches different set of tables to complete the processing.
- perform hash-chaining on hash collision
- Performance benefit in avoiding parser/deparser processing times if packet is not changed between passes.